orchard_iso_pallas.py

```diff
@@ -0,0 +1,176 @@
+
+            numerator_b = (c[7] * self.x * self.x * self.x + c[8] * self.x * self.x + c[9] * self.x + c[10]) * self.y
+            denominator_b = self.x * self.x * self.x + c[11] * self.x * self.x + c[12] * self.x + c[13]
+
+            return orchard_pallas.Point(numerator_a / denominator_a, numerator_b / denominator_b)
+
+    def __init__(self, x, y, is_identity=False):
+        self.x = x
+        self.y = y
+        self.is_identity = is_identity
+
+        if is_identity:
+            assert self.x == Fp.ZERO
+            assert self.y == Fp.ZERO
+        else:
+            assert self.y * self.y == self.x * self.x * self.x + PALLAS_ISO_A * self.x + PALLAS_ISO_B
+
+    def identity():
+        p = Point(Fp.ZERO, Fp.ZERO, True)
+        return p
+
+    def __neg__(self):
+        if self.is_identity:
+            return self
+        else:
+            return Point(Fp(self.x.s), -Fp(self.y.s))
+
+    def __add__(self, a):
+        if self.is_identity:
+            return a
+        elif a.is_identity:
+            return self
+        else:
+            # <https://core.ac.uk/download/pdf/10898289.pdf> section 4.1
+            (x1, y1) = (self.x, self.y)
+            (x2, y2) = (a.x, a.y)
+
+            if x1 == x2:
+                if (y1 != y2) or (y1 == Fp(0)):
+                    return Point.identity()
+                else:
+                    return self.double()
+            else:
+                λ = (y1 - y2) / (x1 - x2)
+                x3 = λ*λ - x1 - x2
+                y3 = λ*(x1 - x3) - y1
+                return Point(x3, y3)
+
+    def __sub__(self, a):
+        return (-a) + self
+
+    def double(self):
+        if self.is_identity:
+            return self
+
+        # <https://core.ac.uk/download/pdf/10898289.pdf> section 4.1
+        λ = (Fp(3) * self.x * self.x + PALLAS_ISO_A) / (self.y + self.y)
+        x3 = λ*λ - self.x - self.x
+        y3 = λ*(self.x - x3) - self.y
+        return Point(x3, y3)
+
+    def __mul__(self, s):
+        s = format(s.s, '0256b')
+        ret = self.ZERO
+        for c in s:
+            ret = ret.double()
+            if int(c):
+                ret = ret + self
+        return ret
+
+    def __bytes__(self):
+        if self.is_identity:
+            return bytes([0] * 32)
+
+        buf = bytes(self.x)
+        if self.y.s % 2 == 1:
+            buf = buf[:31] + bytes([buf[31] | (1 << 7)])
+        return buf
+
+    def __eq__(self, a):
+        if a is None:
+            return False
+        if not (self.is_identity or a.is_identity):
+            return self.x == a.x and self.y == a.y
+        else:
+            return self.is_identity == a.is_identity
+
+    def __str__(self):
+        if self.is_identity:
+            return 'Point(identity)'
+        else:
+            return 'Point(%s, %s)' % (self.x, self.y)
+
+
+Point.ZERO = Point.identity()
+
+x = Fp(2)
+y2 = x * x * x + PALLAS_ISO_A * x + PALLAS_ISO_B
+y = y2.sqrt()
+assert y is not None
+
+Point.GENERATOR = Point(x, y)
```

Comment:
I also wondered this. Technically any non-zero point will do, though.

If we were to "officially" define a generator, it would be iso_map<sup>-1</sup>((-1, 2)).

Reply:
No reason at all, I was just trying to come up with any nonzero point on the curve.

Reply:
Then we should document that this is not a formally-selected common generator for Pallas, or else everyone will assume it is.

Reply:
Documented, thanks!